### ISSUE TITLE
Using React in SPFx Extension Application Customizer Breaks Site Contents

### ISSUE THREAD

**Issue by steevinBradlee** [SharePoint/sp-dev-docs#2365]
#### Category
- [ ] Question
- [ ] Typo
- [x ] Bug
- [ ] Additional article idea

#### Expected or Desired Behavior

I want to add React to an SPFx extension.

#### Observed Behavior

Adding React (importing React and ReactDOM, inserting an element and rendering to this elements) causes the site contents on the site to stop working; only a blank page is loaded with nothing in the body. If I reload the page multiple times the page will sometimes work. Otherwise the page is blank with no errors but after a while the following error appears:

Uncaught Error: Load timeout for modules: react-dom
http://requirejs.org/docs/errors.html#timeout
    at makeError (require-951f856e.js:1)
    at R (require-951f856e.js:1)
    at require-951f856e.js:1

#### Steps to Reproduce

- Create a SharePoint Framework solution and add an application customizer extension to it.
- Import React and ReactDOM (I didn't install them but importing seemed to work off the bat)
-  Create a DOM element and render to it:

`var element = document.createElement('div');
    element.id = 'copifTacoTray';
    document.body.appendChild(element);
    ReactDOM.render(React.createElement("div", null, 'hello'), document.getElementById('copifTacoTray'));`

I noticed that even if I remove these lines and reload the page, the site contents will continue to not load correctly. I noticed that in my console where I'm running gulp serve, the site doesn't work correctly when I see this line in the console: 

Request: '/node_modules/react/dist/react.js'

I'm not sure why but when the site DOES load, this site doesn't appear for the page refresh.

Any help would be greatly appreciated! :)

**Comment by VesaJuvonen** [comment]
What versions are you using? Notice that SPFx and SPO does NOT yet support React 16.

**Comment by steevinBradlee** [comment]
Part of this is probably my inexperience with NPM packages, but I didn't install React or ReactDOM for the project because VS Code Intellisense picked up the packages in my node_modules folder. I'm not sure how/why this would happen considering I never explicitly installed them and they aren't listed as dependencies in my package.json.

I since tried installing both packages, and their typings so my package.json looks like so:

`"dependencies": {
    "@microsoft/decorators": "1.5.1",
    "@microsoft/sp-application-base": "1.5.1",
    "@microsoft/sp-core-library": "1.5.1",
    "@microsoft/sp-dialog": "1.5.1",
    "@types/es6-promise": "0.0.33",
    "@types/react": "^15.6.6",
    "@types/react-dom": "^15.5.6",
    "@types/webpack-env": "1.13.1",
    "react": "^15.6.2",
    "react-dom": "^15.6.2"
  },
  "devDependencies": {
    "@microsoft/sp-build-web": "1.5.1",
    "@microsoft/sp-module-interfaces": "1.5.1",
    "@microsoft/sp-webpart-workbench": "1.5.1",
    "gulp": "~3.9.1",
    "@types/chai": "3.4.34",
    "@types/mocha": "2.2.38",
    "ajv": "~5.2.2"
  }`

**Comment by OliverZeiser** [comment]
This issue exists since the first day of spfx extensions and not only on the sitecontents, but also on the modern recycle bin page. If I remember correctly, the placeholder is undefined. Refreshing multiple times will make it work eventually and the placeholder will be there. I am guessing it is a timing issue.

BUT: The issue only happens, when loading the extension from localhost via gulp serve. I am guessing, that is why there are not that many complains about it.

To reproduce, build an extension that is using react, create a package without using the --ship parameter and deploy it to the appcatalog. Then register the extension on a site so it gets loaded from localhost and start clicking on sitecontents and recycle bin.

**Comment by steevinBradlee** [comment]
Hi @OliverZeiser! Thanks for the info. Does this mean that if I deploy this and have the web part pull its files from the Office 365 CDN, instead of localhost, it should work fine?

**Comment by OliverZeiser** [comment]
That's at least my experience, yes. Give it a try and make a bundle/package with the --ship parameter to include the files :)

**Comment by steevinBradlee** [comment]
Amazing! That worked! Thanks so much @OliverZeiser :D 

I guess I can close this now. 

TLDR: If you want to use React in an extension, serve the files from a CDN, NOT localhost

**Comment by OliverZeiser** [comment]
No Problem. Well it is up to you if you want to close it. The issue is not really resolved :)
